crates/external_services/src/azure_key_vault/core.rs

```diff
@@ -0,0 +1,307 @@
+//! Interactions with the AZURE KEY VAULT SDK
+
+use std::{fmt, sync::Arc, time::Instant};
+
+use azure_identity::DefaultAzureCredential;
+use azure_security_keyvault_keys::{
+    models::{JsonWebKeyEncryptionAlgorithm, KeyOperationsParameters},
+    KeyClient,
+};
+use base64::Engine;
+use common_utils::errors::CustomResult;
+use error_stack::{report, ResultExt};
+use router_env::logger;
+
+use crate::{consts, metrics};
+
+/// Configuration parameters required for constructing a [`AzureKeyVaultClient`].
+#[derive(Clone, Debug, Default, serde::Deserialize)]
+#[serde(default)]
+pub struct AzureKeyVaultConfig {
+    /// key name of Azure Key vault used to encrypt or decrypt data
+    pub key_name: String,
+    /// The Azure vault url of the Key vault.
+    pub vault_url: String,
+    /// version of the key name
+    pub version: String,
+    /// algorithm used to encrypt and decrypt keys
+    pub algorithm: Option<JsonWebKeyEncryptionAlgorithm>,
+}
+
+impl AzureKeyVaultConfig {
+    /// Verifies that the [`AzureKeyVaultClient`] configuration is usable.
+    pub fn validate(&self) -> Result<(), &'static str> {
+        use common_utils::{ext_traits::ConfigExt, fp_utils::when};
+
+        when(self.key_name.is_default_or_empty(), || {
+            Err("Azure Key Vault key name must not be empty")
+        })?;
+
+        when(self.vault_url.is_default_or_empty(), || {
+            Err("Azure Key Vault url must not be empty")
+        })
+    }
+}
+
+/// Client for AZURE KEY VAULT operations.
+#[derive(Clone)]
+pub struct AzureKeyVaultClient {
+    inner_client: Arc<KeyClient>,
+    key_name: String,
+    version: String,
+    algorithm: JsonWebKeyEncryptionAlgorithm,
+}
+
+/// Implement Debug for AzureKeyVaultClient as KeyClient doesn't implement Debug Trait
+impl fmt::Debug for AzureKeyVaultClient {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "AzureKeyVaultClient(inner_client: KeyClient(endpoint: {}), key_name: {}, version: {})",
+            self.inner_client.endpoint(),
+            self.key_name,
+            self.version
+        )
```

Comment:
Please use [`debug_struct()`](https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.debug_struct) instead of manually calling `write!()`.